watchtower-plugin/src/retrier.rs

```diff
@@ -1,171 +1,253 @@
+                            log::info!("Retry strategy succeeded for {}", tower_id);
+                            state.set_tower_status(tower_id, crate::TowerStatus::Reachable);
+                        }
+                        Err(e) => {
+                            log::warn!("Retry strategy gave up for {}. {}", tower_id, e);
+                            // Notice we'll end up here after a permanent error. That is, either after finishing the backoff strategy
+                            // unsuccessfully or by manually raising such an error (like when facing a tower misbehavior)
+                            if let Some(tower) = state.towers.get_mut(&tower_id) {
+                                if tower.status.is_unreachable() {
+                                    log::warn!("Setting {} as unreachable", tower_id);
+                                    state.set_tower_status(
+                                        tower_id,
+                                        crate::TowerStatus::Unreachable,
+                                    );
+                                }
+                            } else {
+                                log::info!("Skipping retrying abandoned tower {}", tower_id);
+                            }
                         }
-                    } else {
-                        log::info!("Skipping retrying abandoned tower {}", tower_id);
                     }
-                }
+                });
             }
         }
     }
 
-    async fn add_appointment(&self, tower_id: TowerId) -> Result<(), Error<&'static str>> {
+    /// Adds an appointment to pending for a given tower.
+    ///
+    /// If the tower is not currently being retried, a new entry for it is created, otherwise, the data is appended to the existing entry.
+    ///
+    /// Returns true if a new entry is created, false otherwise.
+    fn add_pending_appointment(&mut self, tower_id: TowerId, locator: Locator) -> Option<Retrier> {
+        let mut retriers = self.retriers.lock().unwrap();
+        if let std::collections::hash_map::Entry::Vacant(e) = retriers.entry(tower_id) {
+            log::debug!(
+                "Creating a new entry for tower {} with locator {}",
+                tower_id,
+                locator
+            );
+            self.wt_client
+                .lock()
+                .unwrap()
+                .set_tower_status(tower_id, crate::TowerStatus::TemporaryUnreachable);
+
+            let retrier = Retrier::new(self.wt_client.clone(), locator);
+            e.insert(retrier.clone());
+
+            Some(retrier)
+        } else {
+            log::debug!(
+                "Adding pending appointment {} to existing tower {}",
+                locator,
+                tower_id
+            );
+            retriers
+                .get(&tower_id)
+                .unwrap()
+                .pending_appointments
+                .lock()
+                .unwrap()
+                .insert(locator);
```

Comment:
Relating to the race condition above, this insert might get invoked **after** the tower have finished being retried, thus never being retried at all.

Reply:
I think this cannot happen though. Both this method and the `manage_retry` lock `retriers`, so if the tower has finished retrying, the if branch would be hit instead of this one.

In any case, the aforementioned race condition needs fixing.